Add input view component

afx8732 opened the pull request:
This is copied from the original component repo (which is now private) except with extra type-safety and docstrings

Completes #12

afx8732 committed:
Copy input_view from original repo

MeGaGiGaGon commented:
I tried to plug this into #19, and it does display the characters, but my main point of confusion is how it supposed to be used. Mainly, where would the text to type go? I might be using it wrong, but it only displays the characters that have been typed, not the ones that need typing.
My bridge code is in this branch https://github.com/Mannyvv/cj25-monumental-monsteras-team-repo/tree/test-rpg-style-input-with-input-view if you want to take a look at it

afx8732 committed:
Improve docstrings a lot

MeGaGiGaGon commented:
I see, I updated the tie-in branch to work with backspace, and my main concern is that it feels awkward to have both the `Keyboard` and `input_view` store the text state. Ideally it should be in one of the states, probably the `input_view`, since all the current planned input modes don't need access to that information. I think it should be manageable with having something like `input_view.add_char` and `input_view.backspace`.

That could also be further simplified so `input_view` always treats `\N{Symbol for bacspace}` or `\b` as a backspace, but I'm not sure how clear that would be.

afx8732 commented:
The input view doesn't really store the text state internally, it's only actually kept for the scope of the `set_text` method. Once we start making the actual WPM test UI, we'll definitely need actual UI state -- we could put the user input state there, maybe?

MeGaGiGaGon commented:
That sounds reasonable, I forgot about that. Two more things and then I think this should be good to go:
It should better handle the text being too long. 
<img width="445" height="217" alt="image" src="https://github.com/user-attachments/assets/3825b6e4-d1ce-44d2-8b05-be460cda00b7" />
Another reason to centralize the stored text somewhere is that currently `input_view` can't handle the buffer of `Keyboard` overfilling, so having a unified WPM object handle that would fix it. That's something for later though.

It would be nice if the cursor was visible even if nothing has been entered yet
<img width="374" height="54" alt="image" src="https://github.com/user-attachments/assets/9d3aa738-4c59-4aaa-b9cc-d1ebf19e09e0" />

afx8732 committed:
Make cursor visible before first set_text call

afx8732 commented:
The overflow bug doesn't happen for me -- what browser are you using?  
<img width="528" height="258" alt="image" src="https://github.com/user-attachments/assets/9c453d83-5b01-4d1b-bc7a-b20e6547b421" />


Additionally, when parsing the text to make the highlighting, it internally actually just discards extra text since when doing the WPM test, since the test ends on the last key anyways. I could change this if needed, though

MeGaGiGaGon commented:
Brave, but also happens on Opera.
It looks like strange things can happen on other browsers if you shrink the window (this is on Chrome and Edge):
<img width="395" height="224" alt="image" src="https://github.com/user-attachments/assets/d27b7dda-8a04-44bd-9bf9-33164c84313b" />

afx8732 commented:
For me, setting the zoom level to anything but default makes that weird overflow thing happen.

afx8732 committed:
Fixed weird overflow bug

This time instead of using absolute positioning hacks it uses grid hacks to overlay
the foreground over the background

MeGaGiGaGon commented on src/input_view.py:
```diff
@@ -0,0 +1,140 @@
+        with self.text_input:
+            for tok in parsed:
+                ui.label(tok[0]).classes("c" if tok[1] else "w")
+            ui.label("_").classes("cursor")
```
```suggestion
            if len(value) < len(self.full_text):
                ui.label("_").classes("cursor")
```
The WPM manager should handle the case where value reaches the length of the target, but this should stop the size of the input from growing if we don't destroy it instantly on completion.
<img width="375" height="40" alt="image" src="https://github.com/user-attachments/assets/c1639701-b22e-4b25-9b18-fdf6ffcfedcf" />

afx8732 committed:
Hide cursor if input is completed length
